src/components/trail.js

```diff
@@ -72,268 +279,268 @@ AFRAME.registerComponent('trail', {
+    if (this.curvedSegmentsArray.length === this.lifetime) {
+      this.curvedSegmentsArray.shift()
+    }
+    this.curvedSegmentsArray.push(newSegment)
+    return true;
+  },
+
+  createLinearSegment: function (from, to) {
+    return {
+      from: from,
+      amplitude: this.subtractNodes(to, from)
+    }
+  },
+
+  createCurvedSegment: function (handleA, handleB, handleC) {
+    const p00 = this.divideNode(this.sumNodes(handleA, handleB), 2);
+    const p01 = handleB;
+    const p02 = this.divideNode(this.sumNodes(handleB, handleC), 2);
+    const v00 = this.subtractNodes(p01, p00);
+    const v01 = this.subtractNodes(p02, p01);
+    return {
+      p00: p00,
+      p01: p01,
+      v00: v00,
+      v01: v01
+    }
+  },
 
-    this.saberTrajectory.push(sample);
+  calculateRowNodes: function () {
+    const rowNodesArray = [];
+
+    const linearWeight = 0.5;
+    const splinesWeight = this.curvedSegmentsArray.length;
+    const totalWeight = linearWeight + splinesWeight;
+    const linearAmplitude = linearWeight / totalWeight;
+    const splinesAmplitude = splinesWeight / totalWeight;
+
+    let i;
+    let t = 0.0;
+    let localT;
+    const tPerStep = 1 / this.verticalResolution;
+
+    for (i = 0; i < this.rowsCount; i++, t += tPerStep) {
+      if (t <= linearAmplitude) {
+        localT = 1 - t / linearAmplitude;
+        rowNodesArray.push(this.getPointLinear(localT));
+      } else {
+        localT = 1 - (t - linearAmplitude) / splinesAmplitude;
+        rowNodesArray.push(this.getPointSplines(localT));
+      }
+    }
+
+    return rowNodesArray;
+  },
+
+  getPointLinear: function (localT) {
+    const linearSegment = this.linearSegment;
+    return this.sumNodes(linearSegment.from, this.multiplyNode(linearSegment.amplitude, localT))
+  },
+
+  getPointSplines: function (localT) {
+    const tPerSpline = 1 / this.curvedSegmentsArray.length;
+    let splineIndex = Math.floor(localT / tPerSpline);
+    if (splineIndex < 0) splineIndex = 0;
+    if (splineIndex >= this.curvedSegmentsArray.length) splineIndex = this.curvedSegmentsArray.length - 1;
+    const splineT = (localT - tPerSpline * splineIndex) / tPerSpline;
+    return this.evaluateCurvedSegment(this.curvedSegmentsArray[splineIndex], splineT);
+  },
+
+  evaluateCurvedSegment: function (segment, t) {
+    const p10 = this.sumNodes(segment.p00, this.multiplyNode(segment.v00, t));
+    const p11 = this.sumNodes(segment.p01, this.multiplyNode(segment.v01, t));
+    const v10 = this.subtractNodes(p11, p10);
+    return this.sumNodes(p10, this.multiplyNode(v10, t));
+  },
+
+  lerpNode: function (node, t) {
+    return new THREE.Vector3(
+      node.from.x + (node.to.x - node.from.x) * t,
+      node.from.y + (node.to.y - node.from.y) * t,
+      node.from.z + (node.to.z - node.from.z) * t,
+    );
+  },
 
-    this.initGeometry();
-  }
+  sumNodes: function (nodeA, nodeB) {
+    return {
+      from: new THREE.Vector3(
+        nodeA.from.x + nodeB.from.x,
```

Comment:
Why not use standard Vector3 methods? https://threejs.org/docs/?q=Vector#api/en/math/Vector3